### PR TITLE
calcCCD and calcDeltaCCD now throw errors for genes with zero groupwise variance and require all reference genes to be in emat.

### PR DESCRIPTION
**Pull request by ehoutland**

### CODE REVIEW

File: R/deltaccd.R [hunk 1]
```diff
@@ -1,12 +1,12 @@
 #' @importFrom foreach foreach %do% %dopar%
 #' @importFrom doRNG %dorng%
-#' @importFrom data.table data.table :=
+#' @importFrom data.table data.table as.data.table setnames :=
```
Comment (jakejh) [inline R/deltaccd.R]: If we're only using these functions once, I prefer to not import them and instead use the ::.

File: R/deltaccd.R [hunk 2]
```diff
@@ -152,33 +152,46 @@ calcCCD = function(
     stop('refCor must be a correlation matrix, with identical rownames and colnames.')}
 
   geneNames = rownames(refCor)[rownames(refCor) %in% rownames(emat)]
-  refNow = refCor[geneNames, geneNames]
-  if (length(geneNames) < 2) {
-    stop('Fewer than two genes in the reference are in the expression matrix.')
-  } else if (length(geneNames) < nrow(refCor)) {
-    warning(sprintf('%d gene(s) in reference is/are not in the expression matrix.',
-                    nrow(refCor) - length(geneNames)))}
+  if (length(geneNames) < nrow(refCor)) {
```
Comment (jakejh) [inline R/deltaccd.R]: Let's put this in a separate function, so we can minimize redundancy in calcCCD and calcDeltaCCD. Also we should print out the actual genes, not just how many there are.

File: R/deltaccd.R [hunk 3]
```diff
@@ -152,33 +152,46 @@ calcCCD = function(
...
+                 nrow(refCor) - length(geneNames)))}
 
   if (is.null(groupVec)) {
     groupVec = rep('all', ncol(emat))
   } else if (length(groupVec) != ncol(emat)) {
     stop('Length of groupVec does not match the number of columns in emat.')
   } else if (min(table(groupVec)) < 3) {
     stop('Each unique group in groupVec must have at least three samples.')}
+  
+  varCheck = foreach (groupNow = sort(unique(groupVec)), .combine = rbind) %do% {
```
Comment (jakejh) [inline R/deltaccd.R]: Duplicated code is a good indication of the need for a function.

File: R/deltaccd.R [hunk 4]
```diff
@@ -152,33 +152,46 @@ calcCCD = function(
...
+    setnames(varDt, 'varMat', 'variance')
+    varDt[, group := groupNow]
+    
+    zeroVar = varDt[variance == 0]
+    
+    return(zeroVar)}
+  
+  if (nrow(varCheck) > 0) {
+    stop('Zero variance in the following gene-group pairs:\n', 
+         paste(utils::capture.output(print(varCheck)), collapse = '\n'))}
```
Comment (jakejh) [inline R/deltaccd.R]: If all rows in varCheck have zero variance, why print the variance column?

File: R/deltaccd_plot.R [hunk 5]
```diff
@@ -142,10 +142,12 @@ plotHeatmap = function(geneNames, emat, groupVec = NULL) {
 plotRefHeatmap = function(refCor) {
   if (any(rownames(refCor) != colnames(refCor)) || !isSymmetric(refCor)) {
     stop('refCor must be a correlation matrix, with identical rownames and colnames.')}
+  
+  geneNames = rownames(refCor)
 
   dt = data.table(refCor, gene1 = geneNames)
   dt = data.table::melt(
-    dt, id.vars = 'gene1', measure.vars = 'gene2', value.name = 'rho')
+    dt, id.vars = 'gene1', measure.vars = geneNames, value.name = 'rho')
```
Comment (jakejh) [inline R/deltaccd_plot.R]: This seems like an unrelated change, and I don't see how the new code will work, since line 151 uses a column gene2 that doesn't seem to exist. Have you tested it?

**Review by jakejh — CHANGES_REQUESTED** [review]

File: R/deltaccd_plot.R (hunk 5 above)
Comment (jakejh) [reply R/deltaccd_plot.R]: I repeat my question.

File: R/deltaccd.R (hunk 3 above)
Comment (jakejh) [reply R/deltaccd.R]: There's still duplicated code in the `if` and `stop`. Where should that code go?

**Comment by jakejh** [comment]
Let's also add a separate function to check `refCor`.

File: R/deltaccd.R [hunk 6]
```diff
@@ -87,6 +87,33 @@ calcCCDSimple = function(ref, emat, method = 'spearman', scale = FALSE) {
...
+  return(varCheck)}
+
+
+checkGenes = function(emat, refCor) {
+  geneNames = rownames(refCor)[rownames(refCor) %in% rownames(emat)]
+  if (length(geneNames) < nrow(refCor)) {
+    missingGenes = setdiff(rownames(refCor), geneNames)
+    stop(paste0('The following gene(s) is/are not in the expression matrix: \n',
+                paste0(missingGenes, collapse = '\n')))
+    } else { return(geneNames) }}
```
Comment (jakejh) [inline R/deltaccd.R]: The `return` statement doesn't need to be inside an `else`, do you see why?

File: R/deltaccd_plot.R (hunk 5 above)
Comment (ehoutland) [reply R/deltaccd_plot.R]: Definitely a typo.

File: R/deltaccd.R (hunk 6 above)
Comment (ehoutland) [reply R/deltaccd.R]: Yep, if the `if` condition is met, then the function will throw an error before `geneNames` is returned anyway.

File: R/deltaccd_plot.R (hunk 5 above)
Comment (jakejh) [reply R/deltaccd_plot.R]: That doesn't answer my question, though. Does the function work as is? Have you tested it? Is there a gene2 column?

File: R/deltaccd.R (hunk 3 above)
Comment (ehoutland) [reply R/deltaccd.R]: It probably would make sense to put it in the `checkVar` function.

File: R/deltaccd_plot.R (hunk 5 above)
Comment (ehoutland) [reply R/deltaccd_plot.R]: No, it wasn't working. I meant to type `variable.names = 'gene2'`, not `measure.vars`. Also, I needed to change `ematNow` to `refCor`.

File: R/deltaccd.R [hunk 7]
```diff
@@ -87,6 +87,45 @@ calcCCDSimple = function(ref, emat, method = 'spearman', scale = FALSE) {
...
+  geneNames = rownames(refCor)[rownames(refCor) %in% rownames(emat)]
+  if (length(geneNames) < nrow(refCor)) {
+    missingGenes = setdiff(rownames(refCor), geneNames)
+    stop(paste0('The following gene(s) is/are not in the expression matrix: \n',
+                paste0(missingGenes, collapse = '\n')))} 
+  
+  return(geneNames)}
+
+
+checkRefCor = function(refCor, refEmat = NULL, method = 'spearman') {
```
Comment (jakejh) [inline R/deltaccd.R]: Look at the code more carefully. What value should the function return? This is why we need tests.

Comment (ehoutland) [reply R/deltaccd.R]: Oops, I think it needs to return `refCor` for the rest of calcCCD/calcDeltaCCD to work with a `refEmat`.

File: R/deltaccd.R [hunk 8]
```diff
@@ -197,7 +195,7 @@ calcCCD = function(
   } else if (min(table(groupVec)) < 3) {
     stop('Each unique group in groupVec must have at least three samples.')}
   
-  checkVar(emat[geneNames, ], groupVec)
+  checkVar(emat, groupVec)
```
Comment (jakejh) [inline R/deltaccd.R]: Why not make the logic here depend on the value of nPerm?

**Comment by jakejh** [comment]
The last step is to bump the version in the DESCRIPTION, update the NEWS.md file, then run `pkgdown::build_site()`.

**Comment by jakejh** [comment]
I meant update the version of the package itself, not the versions of dependencies. There are some conflicts, so you'll need to merge master into your local zero_var branch first, resolve conflicts in the NEWS and code (make sure to remove `'geneNames'` from the global variables), rerun `pkgdown::build_site`, and push the commits so we can finish the pull request.

File: R/deltaccd.R [hunk 9]
```diff
@@ -5,7 +5,8 @@ NULL
 
 
 globalVariables(c(
-  'i', 'groupNow', 'group', '.', 'gene1', 'gene2', 'rho', 'group2Now', '.SD'))
+  'i', 'groupNow', 'group', '.', 'gene1', 'gene2', 'rho', 'group2Now',
+  'ematNow', '.SD', 'variance'))
```
Comment (jakejh) [inline R/deltaccd.R]: Are you sure ematNow needs to be treated like a global variable?

Comment (ehoutland) [reply R/deltaccd.R]: `devtools::check` says no, but then I guess I don't really understand when you _do_ need to declare a global variable?

Comment (jakejh) [reply R/deltaccd.R]: Ideally never, but in practice we do it for columns that get treated like variables by data.table or iterated variables in a foreach loop, since R's package check doesn't realize those aren't actually global variables.
